src/protocol/serial.rs

```diff
@@ -0,0 +1,215 @@
+                Status::Success => Ok(()),
+                e => Err(e),
+            }
+        }
+    }
+
+    /// Set the serial device's attributes. Passing `None` to any argument will keep the device
+    /// default. Note that each attribute is set each time this function is called, so setting one
+    /// attribute in one call and another in the next will result in the first attribute being
+    /// reset to the device default.
+    pub fn set_attributes(
+        &self,
+        baud_rate: Option<u64>,
+        receive_fifo_depth: Option<u32>,
+        timeout: Option<u32>,
+        parity: Option<ParityType>,
+        data_bits: Option<u8>,
+        stop_bits: Option<StopBits>,
+    ) -> Status {
+        unsafe {
+            (self.set_attributes)(
+                self,
+                baud_rate.unwrap_or(0),
+                receive_fifo_depth.unwrap_or(0),
+                timeout.unwrap_or(0),
+                parity.unwrap_or(ParityType::DefaultParity),
+                data_bits.unwrap_or(0),
+                stop_bits.unwrap_or(StopBits::DefaultStopBits),
+            )
+        }
+    }
+
+    /// Set the serial device's control bits. Per the UEFI specification, only five of the control
+    /// bits can be set with SetControlBits:
+    ///
+    /// * `REQUEST_TO_SEND`
+    /// * `DATA_TERMINAL_READY`
+    /// * `HARDWARE_LOOPBACK_ENABLE`
+    /// * `SOFTWARE_LOOPBACK_ENABLE`
+    /// * `HARDWARE_FLOW_CONTROL_ENABLE`
+    pub fn set_control_bits(&self, control: ControlBits) -> Result<(), Status> {
+        unsafe {
+            match (self.set_control_bits)(self, control.bits()) {
+                Status::Success => Ok(()),
+                e => Err(e),
+            }
+        }
+    }
+
+    /// Retrieve the serial device's control bits.
+    pub fn get_control_bits(&self) -> Result<ControlBits, Status> {
+        let mut bits: u32 = 0;
+        unsafe {
+            match (self.get_control_bits)(self, &mut bits) {
+                Status::Success => Ok(ControlBits::from_bits_truncate(bits)),
+                e => Err(e),
+            }
+        }
+    }
+
+    /// Write some data to the serial device.
+    pub fn write_raw(&self, data: *const u8, length: usize) -> Result<usize, Status> {
+        let mut new_length = length.clone();
+        unsafe {
+            match (self.write)(self, &mut new_length, data as *const CVoid) {
+                Status::Success => Ok(new_length),
+                e => Err(e),
+            }
+        }
+    }
+
+    /// Write a string to the serial device.
+    pub fn write(&self, data: &str) -> Result<usize, Status> {
+        self.write_raw(data.as_ptr(), data.len())
+    }
+
+    fn read_raw(&self, buf: *mut u8, length: usize) -> Result<Option<(*const u8, usize)>, Status> {
+        let mut new_length = length.clone();
+        unsafe {
+            match (self.read)(self, &mut new_length, buf as *mut CVoid) {
+                Status::Success => Ok(Some((buf, new_length))),
```

Comment:
I assume you're returning `buf` for chaining purposes?  If so, is that a sufficient reason to do it?